designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,209 @@
+---
+title: controller-observation-plugin
+authors:
+  - "@Kavinjsir"
+reviewers:
+  - "@varsha"
+  - "@camila"
+approvers:
+creation-date: 2022-06-06
+last-updated: 2022-06-06
+status: provisional
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# a-grafana-dashboard-plugin-to-observe-controller-metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+- TBD
+
+## Summary
+
+This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.
+
+The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.
+
+In the context of this EP, the Grafana 'manifests' are mainly:
+
+- JSON files to be directly loaded on Grafana dashboards
+- TBD
+
+## Motivation
+
+The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.
+
+Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
+It would be great if we can provide a unified and comprehensive approach so that users may:
+
+- easily visualize controller/operator metrics
+- avoid duplicate creation of Grafana dashboards file
+- improve the observability towards the opeartor, keep it safe and robust
+
+### Goals
+
+- Provide manifests that can display operator status for common needs.
+- Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.
+
+### Non-Goals
+
+- TBD
+
+## Proposal
+
+This proposed implementation is to create a new plugin, `grafana`, that scaffolds the manifests which are loaded by Grafana to display dashboards for controllers' status.
+
+Basically, the manifest can be raw json file such as `controller-dashboard.json` that can be directly applied to Grafana Web UI.
+
+As there are many useful metrics, it might be better to observe in multiple dashboards rather than one.
+
+Hence, we may also provide optional dashboards that focus on different aspects.
+
+For instance: work queue, reconciliation, performance, webhook.
+
+As a result, the layout given by the plugin can be a folder of `config/grafana/`:
```

Comment:
In the config/ dir we have only manifests to be used with kustomize. See that SDK add new features by adding manifests there, i.e for scorecard (https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/config/scorecard) however, all is using kustomize, so I am afraid that we should not add JSON files on the config/ dir. However, we can create a new directory in the root like grafana/ where we can add all there.

Reply:
Yeah, this sounds better!